From c7ea03b11757acfd1eb22eac8750359b3eb680c7 Mon Sep 17 00:00:00 2001
From: Hans Boehm <Hans.Boehm@hp.com>
Date: Thu, 15 Mar 2012 21:09:05 +0400
Subject: [PATCH 3/4] Speedup calloc size overflow check by preventing
 division if small values

* malloc.c (GC_SQRT_SIZE_MAX): New macro.
* malloc.c (calloc): Add fast initial size overflow check to avoid
integer division for reasonably small values passed.
---
 malloc.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/malloc.c b/malloc.c
index be6d241..a2aa41e 100644
--- a/malloc.c
+++ b/malloc.c
@@ -354,9 +354,12 @@ void * malloc(size_t lb)
 # define GC_SIZE_MAX (~(size_t)0)
 #endif
 
+#define GC_SQRT_SIZE_MAX ((1U << (WORDSZ / 2)) - 1)
+
 void * calloc(size_t n, size_t lb)
 {
-    if (lb && n > GC_SIZE_MAX / lb)
+    if ((lb | n) > GC_SQRT_SIZE_MAX /* fast initial test */
+        && lb && n > GC_SIZE_MAX / lb)
       return NULL;
 #   if defined(GC_LINUX_THREADS) /* && !defined(USE_PROC_FOR_LIBRARIES) */
 	/* libpthread allocated some memory that is only pointed to by	*/
-- 
1.7.10.2

